src/include/mpir_refcount_pobj.h

```diff
@@ -10,45 +10,24 @@
 /* define a type for the completion counter */
 #include "opa_primitives.h"
 
-typedef OPA_int_t MPIR_cc_t;
+typedef MPL_atomic_int_t MPIR_cc_t;
 
-/* implies no barrier, since this routine should only be used for request
- * initialization */
 static inline void MPIR_cc_set(MPIR_cc_t * cc_ptr, int val)
 {
-    if (val == 0) {
-        /* values other than 0 do not enforce any ordering, and therefore do not
-         * start a HB arc */
-        /* MT FIXME using cc_set in this way is sloppy.  Sometimes the caller
-         * really does know that the cc value may cleared, but more likely this
-         * is just a hack to avoid the work of figuring out what the cc value
-         * currently is and decrementing it instead. */
-        /* barrier ensures that any state written before indicating completion is
-         * seen by the thread polling on the cc.  If OPA adds store-release
-         * semantics, we can convert to that instead. */
-        OPA_write_barrier();
-        MPL_VG_ANNOTATE_HAPPENS_BEFORE(cc_ptr);
-    }
-#if defined(MPL_VG_AVAILABLE)
-    /* MT subtle: store_int is actually safe to use, but Helgrind/DRD/TSan all
-     * view the store/load pair as a race.  Using an atomic operation for the
-     * store side makes all three happy.  DRD & TSan also support
-     * ANNOTATE_BENIGN_RACE, but Helgrind does not. */
-    OPA_swap_int(cc_ptr, val);
-#else
-    OPA_store_int(cc_ptr, val);
-#endif
+    /* Using relaxed semantics since this routine should only be used for
+     * request initialization. */
+    MPL_atomic_relaxed_store_int(cc_ptr, val);
```

Comment:
This is used to "set" counters, so it should guarantee the acquire-release semantics (e.g., `MPIDI_UCX_recv()`)
https://github.com/pmodels/mpich/blob/master/src/mpid/ch4/netmod/ucx/ucx_recv.h#L53

This is necessary for your commit `ch4/ucx: remove OPA_read_write_barrier`.

`MPL_atomic_(release_)store_int` is okay.

Reply:
During request initialization, the request ptr is not visible to other threads, so no contention will occur at "set". The request ptr is returned to the same thread and has to go through some critical section before other thread sees this request ptr and be able to access its content. So I believe that critical section for sharing the request ptr itself makes "relaxed" setting `cc_ptr` here ok.  What do you think?

Reply:
For initialization, relaxed is okay, but I think this set function is used for synchronization as well.
For example, `MPIDI_UCX_recv_cmpl_cb()` relies on its acquire/release semantics and (as far as I remember) that's why we can remove `OPA_read_write_barrier()` in one of your commits (`ch4/ucx: remove OPA_read_write_barrier`)
https://github.com/pmodels/mpich/blob/master/src/mpid/ch4/netmod/ucx/ucx_recv.h#L53

I might not understand the algorithm well. I thought thread synchronization is done by this `rreq->cc`, but is it actually realized by availability of a request pointer (e.g., `rreq` is pushed to somewhere else in a release-store way and other threads can read it in an acquire-load way)? If this is correct, `MPIR_cc_set()` is just an initialization and does not require release-store. (`MPIR_cc_init()` would be a better name.)

Reply:
> For initialization, relaxed is okay, but I think this set function is used for synchronization as well.

`set` is only used during initialization. Synchronization is done exclusively through atomic `inc/dec`. 

> I might not understand the algorithm well. I thought thread synchronization is done by this `rreq->cc`, but is it actually realized by availability of a request pointer (e.g., `rreq` is pushed to somewhere else in a release-store way and other threads can read it in an acquire-load way)? If this is correct, `MPIR_cc_set()` is just an initialization and does not require release-store. (`MPIR_cc_init()` would be a better name.)

`rreq` itself is not atomic, but it always go through some critical section before it is returned and made available to other threads. For example, in a typical operation, `MPI_Irecv` may create and initialize a request object (this maybe outside critical section but thread-safe), then it will enter a critical section to post to the `recv-queue`, then exit the critical section before returning the request to the user after which the user may invoke progress wait on its "completion counter" in a different threads. My question is, does this critical section in-between the initialization and the actual usage of completion counter synchronization ensures correctness?

Reply:
The request creation is in a performance critical part so any cycles we can save is good.

Reply:
Tagging @rzambre for potential interest.

Reply:
> My question is, does this critical section in-between the initialization and the actual usage of completion counter for synchronization ensures correctness?
It means that,
- the critical section is used to synchronize threads when `cc = 0`, and
- the completion counter (`cc`) is used to synchronize threads when `cc > 0`.
If so, it truly ensures correctness as you wrote.

I still doubt this assumption, though. Do all the completion counter use cases follow this model? 

Just a confirmation: why does `MPIR_cc_is_complete()` use `MPL_atomic_load_int()` (i.e., acquire load)?
Intuitively, if `cc` is set to zero by relaxed store, 
- in terms of release-acquire semantics, this relaxed store won't be visible to `MPL_atomic_load_int()` without synchronization via the critical section.
-if `MPIR_cc_is_complete()` is always done after the critical section synchronization, do we really need `MPL_atomic_load_int()` in `MPIR_cc_is_complete()`?
Since this part looks tricky and hard to debug in the future, I want to clarify this more.

Reply:
> 
> 
> > My question is, does this critical section in-between the initialization and the actual usage of completion counter for synchronization ensures correctness?
> 
> It means that,
> 
>     * the critical section is used to synchronize threads when `cc = 0`, and
>
Before and within the critical section, only the creation thread is setting `cc`, so does it matter that `cc=0`? In fact, we always set it greater than 0, indicating number of components that need complete before the whole request is completed.

>     * the completion counter (`cc`) is used to synchronize threads when `cc > 0`.
> 
Yes, always.
> 
> If so, it truly ensures correctness as you wrote.
> 
> I doubt this assumption, though. Do all the completion counter use cases follow this model?

Yes, always.
> 
> Just a confirmation: why does `MPIR_cc_is_complete()` use `MPL_atomic_load_int()` (i.e., acquire load)? Intuitively, if `cc` is set to zero by relaxed store,
> 
>     * in terms of release-acquire semantics, this relaxed store won't be visible to `MPL_atomic_load_int()` without synchronization via the critical section, or

But it (the relaxed store) will be visible after the first critical section, right? `MPIR_cc_is_complete()` is only used after the first critical section.
> 
>     * if `MPIR_cc_is_complete()` is always done after the critical section synchronization, do we really need `MPL_atomic_load_int()` in `MPIR_cc_is_complete()`?
> 
After the first critical section (that init and enqueues the request), potentially different threads will be waiting on the `cc` counter (polling `MPIR_cc_is_complete`) and completes (decrement) the `cc` counter.
> 
> Since this part looks tricky and hard to debug in the future, I want to clarify this more.

That's what I am counting on you :)

Reply:
I read the code, and it seems that this `set` is used only for initialization. `relaxed` should be sufficient, then. Thanks for clarification, @hzhou!